### PR TITLE
AP_Math: added comments and a test for euler ordering

### PR DESCRIPTION
**Pull request by tridge**
our main euler functions did not have a comment on the ordering convention

### CODE REVIEW

**Review by rmackay9 — APPROVED** [review]

File: libraries/AP_Math/matrix3.h [hunk 1]
```diff
@@ -235,13 +235,17 @@ class Matrix3 {
         return a.is_nan() || b.is_nan() || c.is_nan();
     }
 
-    // create a rotation matrix from Euler angles
+    /*
+      create a rotation matrix from Euler angles in 321 euler orderin
```
Comment (peterbarker) [inline libraries/AP_Math/matrix3.h:239]:
```suggestion
      create a rotation matrix from Euler angles in 321 euler ordering
```

**Review by peterbarker — APPROVED** [review]

**Comment by andyp1per** [comment]
Having scratched my head against this for some time it would be great to see the additional clarifying commentary:

- Say whether rotations are intrinsic or extrinsic? (https://github.com/ArduPilot/ardupilot/pull/27335)
- Say what the positive direction is for each axis
- This doc in matrix3.cpp now gives a 404 - http://gentlenav.googlecode.com/files/EulerAngles.pdf

**Review by andyp1per — COMMENTED** [review]
I think its great as far as it goes, but possible could do with some additional commentary

**Comment by isso11** [comment]
> * positive direction is for each axis
I think so too. In the context of Euler angles and the North-East-Down (NED) coordinates frame:
Yaw (ψ): Positive is clockwise when viewed from above, around the Down (Z) axis.
Pitch (θ): Positive is nose up, around the East (Y) axis.
Roll (𝜙): Positive is right wing down, around the North (X) axis.

**Comment by andyp1per** [comment]
@tridge I manually checked one of these and either euler 321 means something different to what I think it does or these rotations are euler 123. An easy example take ROTATION_ROLL_90_YAW_90, this means roll 90 followed by yaw 90. The code swaps around vector values so that [x, y, z] -> [z, x, y] which indeed is roll 90 followed by yaw 90 it is *not* yaw 90 followed by roll 90. So either the meaning of euler 321 is very weird - which I am prepared to believe - or your documentation is actually wrong.

**Review by andyp1per — CHANGES_REQUESTED** [review]
Need to resolve my very confused understanding of the meaning of euler 321

**Comment by isso11** [comment]
> euler 321

Aren't these rotations basically for transformations between body fixed and local inertial NED frames? If so, then the rotations 321 should, by standard definition, mean  [Yaw (3, around the 3rd axis, Z), Pitch (2, around the 2nd axis, Y), Roll(1, around the 1st axis, X)] in that order.

**Comment by andyp1per** [comment]
> Aren't these rotations basically for transformations between body fixed and local inertial NED frames? If so, then the rotations 321 should, by standard definition, mean [Yaw (3, around the 3rd axis, Z), Pitch (2, around the 2nd axis, Y), Roll(1, around the 1st axis, X)] in that order.

That is my understanding

**Comment by Georacer** [comment]
> That is my understanding

Mine as well.

**Comment by tridge** [comment]
@andyp1per @Georacer see the rotate_ordered() function in this PR

**Comment by tridge** [comment]
> * Say whether rotations are intrinsic or extrinsic?

as demonstrated by the rotate_ordered() function it is intrinsic

**Comment by tridge** [comment]
> * This doc in matrix3.cpp now gives a 404

it is common that URLs like this disappear. If we can find the PDF we could put it in https://github.com/ArduPilot/Datasheets

**Comment by andyp1per** [comment]
@tridge I have pushed a test that fails to prove my point.

File: libraries/AP_Math/matrix3.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_Math/matrix3.h]: changed

**Comment by tridge** [comment]
@andyp1per I've found the original EulerAngles.pdf from Bill Premerlani and added it to out Datasheets repo and updated the link in the code in this PR

**Comment by peterbarker** [comment]
Ping @andyp1per can we merge this one?

**Review by andyp1per — APPROVED** [review]

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**